src/libopensc/iso7816.c

```diff
@@ -165,31 +165,113 @@ iso7816_read_binary(struct sc_card *card, unsigned int idx, u8 *buf, size_t coun
 }
 
 
+const struct sc_asn1_entry c_asn1_do_data[] = {
+	{ "Offset Data Object", SC_ASN1_OCTET_STRING, SC_ASN1_APP|0x14, SC_ASN1_OPTIONAL, NULL, NULL },
+	{ "Discretionary Data Object", SC_ASN1_OCTET_STRING, SC_ASN1_APP|0x13, SC_ASN1_ALLOC|SC_ASN1_UNSIGNED, NULL, NULL },
+	{ NULL, 0, 0, 0, NULL, NULL }
+};
+
+int encode_do_data(struct sc_context *ctx,
+		unsigned int idx, const unsigned char *data, size_t data_len,
+		u8 **out, size_t *outlen)
+{
+	unsigned char offset_buffer[2];
+	size_t offset_buffer_len = sizeof offset_buffer;
+	struct sc_asn1_entry asn1_do_data[sizeof c_asn1_do_data / sizeof *c_asn1_do_data];
+	sc_copy_asn1_entry(c_asn1_do_data, asn1_do_data);
+
+	if (idx > 0xFFFF)
+		LOG_TEST_RET(ctx, SC_ERROR_INTERNAL, "Offset beyond 0xFFFF not supported");
+	offset_buffer[0] = (u8) (idx >> 8);
+	offset_buffer[1] = (u8) (idx & 0x00FF);
+	sc_format_asn1_entry(asn1_do_data + 0, offset_buffer, &offset_buffer_len, 1);
+
+	if (data && data_len) {
+		sc_format_asn1_entry(asn1_do_data + 1, &data, &data_len, 1);
+	} else {
+		sc_format_asn1_entry(asn1_do_data + 1, NULL, NULL, 0);
+	}
+
+	int r = sc_asn1_encode(ctx, asn1_do_data, out, outlen);
```

Comment:
Some compilers will have issues with interleaved definitions and declaration. Please, move the `int r` to the beginning of the function.

Reply:
Same in the following function `decode_do_data()`

Reply:
done